File: courses/codes/CompCan.java
import java.util.*;

public class CompCan {
    // Method to print matrix
    public static void printMatrix(int[][] M, int ov, int nv) {
        for (int i = 0; i < M.length; i++) {
            for (int j = 0; j < M[0].length; j++) {
                int v = (M[i][j] == ov) ? nv : M[i][j];
                System.out.print(v + " ");
            }
            System.out.println();
        }
    }

    // Floyd-Warshall Algorithm
    public static int[][] floydWarshall(int[][] W) {
        int n = W.length;
        int INF = Integer.MAX_VALUE / 2; // Avoid overflow
        
        int[][] D = new int[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(D[i], INF);
        }

        // Copy initial weights
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (W[i][j] > 0) {
                    D[i][j] = W[i][j];
                }
            }
        }

        for (int i = 0; i < n; i++) {
            D[i][i] = 0;
        }

        // Floyd-Warshall logic
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (D[i][k] < INF && D[k][j] < INF) {
                        D[i][j] = Math.min(D[i][j], D[i][k] + D[k][j]);
                    }
                }
            }
        }
        
        return D;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int N = scanner.nextInt();
        int M = scanner.nextInt();

        int[][] W = new int[N][N];

        for (int i = 0; i < M; i++) {
            int from = scanner.nextInt() - 1;
            int to = scanner.nextInt() - 1;
            int weight = scanner.nextInt();
            W[from][to] = weight;
        }

        int[][] D = floydWarshall(W);
        printMatrix(D, Integer.MAX_VALUE / 2, -1);
        
        scanner.close();
    }
}

File: courses/codes/cclv1-sr-class/compcan.py
'''
    Print Matrix M
    Change value ov to new value nv if the parameters are supplied
'''
def print_matrix(M, ov='', nv=''):
    if ov == '' and nv == '' or ov == nv:
        for i in range(len(M)):
            for j in range(len(M[0])):
                print(int(M[i][j]), end=' ')
            print()
    else:
        for i in range(len(M)):
            for j in range(len(M[0])):
                v = nv if M[i][j] == ov else M[i][j]
                print(int(v), end=' ')
            print()


'''
    Applies Floyd-Warshall Algorithm on the adj matrix of weights W
'''
def floyd_warshall(W):
    n = len(W)

    D = [[float('inf')] * n for _ in range(n)]

    for i in range(n):
        for j in range(n):
            if W[i][j] > 0:
                D[i][j] = W[i][j]

    for i in range(n):
        D[i][i] = 0

    for k in range(n):
        for i in range(n):
            for j in range(n):
                D[i][j] = min(D[i][j], D[i][k] + D[k][j])

    return D

N = int(input())
M = int(input())

W = [[0 for x in range(N)] for _ in range(N)]

for _ in range(M):
    f, t, w = tuple(map(int, input().split()))
    W[f - 1][t - 1] = w

D = floyd_warshall(W)
print_matrix(D, float('inf'), -1)
